rails/database-migrations.md

```diff
@@ -0,0 +1,138 @@
+
+Other common gotchas to test for:
+- The migration won't run on a fresh database because there was something in the local dev db it was relying on.
+- Use realistic production sized data to check for speed.
+
+## Use Cases
+
+### Use ActiveRecord associations in a migration
+
+Existing model:
+
+```ruby
+class Tag
+  belongs_to :blog_post
+end
+
+class BlogPost
+  has_many :tags
+end
+```
+
+You decide for performance reasons to move the tags into an array column on the `blog_posts` table instead of having them in a separate `tags` table.
+
+Write the migration this way:
+
+```ruby
+class MoveTags < ActiveRecord::Migration[5.2]
+  # Inherit from ActiveRecord::Base instead of ApplicationRecord since the
+  # ApplicationRecord is subject to changes as the app changes.
+  class Tag < ActiveRecord::Base; end
+
+  class BlogPost < ActiveRecord::Base
+    has_many :tags
+  end
+
+  def up
+    add_column :blog_posts, :categories, :string, array: true
+
+    say("Moving #{Tag.count} tags to #{BlogPost.count} blog posts")
+
+    # Use find_each to minimize memory usage when dealing with many records
+    BlogPost.find_each(batch_size: 100) do |blog_post|
+      tags = blog_post.tags.pluck(:value)
+
+      # You can never have too much logging when migrating data!
+      say("Moving tags #{tags} to BlogPost #{blog_post.id}")
+
+      blog_post.update!(categories: tags)
+    end
+  end
+
+  def down
+    remove_column :blog_posts, :categories
+  end
+end
+```
+
+**Output:**
+```
+$ bin/rails db:migrate
+== 20190604160527 MoveTags: migrating =========================================
+-- add_column(:blog_posts, :categories, :string, {:array=>true})
+   -> 0.0006s
+-- Moving 6 tags to 3 blog posts
+-- Moving tags ["tag-0", "tag-1"] to BlogPost 5
+-- Moving tags ["tag-0", "tag-1"] to BlogPost 6
+-- Moving tags ["tag-0", "tag-1"] to BlogPost 7
+== 20190604160527 MoveTags: migrated (0.0259s) ================================
+```
+
+A few things to note about this data migration:
+- The `MoveTags` migration defines to child classes: `MoveTags::Tag` and `MoveTags::BlogPost`. These classes are completely different than the classes residing in `app/models`. You can even delete the now unused `Tag` class in `app/models` and this migration will still run!
+- For brevity, this example doesn't worry about dropping the `tags` table, and the `down` method should be expanded to move the tags from blog posts back to tags.
+- Logging is your friend when doing data migrations. Do it early and do it often.
+
+### Heroku review app setup
+For Heroku review apps, use the `postdeploy` script to load the database schema and seed the DB with example data. Use the `release` phase in the `Procfile` to run migrations. This is because `postdeploy` _only runs once_ when the review app is created, whereas the `release` phase _runs on every deploy_ to the review app.
```

Comment:
`release` runs before `postdeploy`, so it doesn't work to do `db:schema:load` in `postdeploy` (see [Heroku's documentation](https://devcenter.heroku.com/articles/release-phase#review-apps-and-the-postdeploy-script)).

`release` needs to run either `db:schema:load` or `db:migrate`, based on if the database is already set up. One example of how to do this: https://gist.github.com/stevenharman/98576bf49b050b9e59fb26626b7cceff.